internal/services/portal/portal_dashboard_data_source.go

```diff
@@ -46,31 +55,83 @@ func dataSourcePortalDashboard() *pluginsdk.Resource {
 func dataSourcePortalDashboardRead(d *pluginsdk.ResourceData, meta interface{}) error {
 	client := meta.(*clients.Client).Portal.DashboardsClient
 	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
-	name := d.Get("name").(string)
-	resourceGroup := d.Get("resource_group_name").(string)
 	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
 	defer cancel()
 
-	id := parse.NewDashboardID(subscriptionId, resourceGroup, name)
-	resp, err := client.Get(ctx, id.ResourceGroup, id.Name)
-	if err != nil {
-		if utils.ResponseWasNotFound(resp.Response) {
-			return fmt.Errorf("portal dashboard %q was not found in Resource Group %q", id.Name, id.ResourceGroup)
+	name := d.Get("name").(string)
+	displayName, displayNameOk := d.GetOk("display_name")
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	var dashboard portal.Dashboard
+
+	if !displayNameOk {
+		var err error
+		dashboard, err = client.Get(ctx, resourceGroup, name)
+		if err != nil {
+			if utils.ResponseWasNotFound(dashboard.Response) {
+				return fmt.Errorf("portal Dashboard %q was not found in Resource Group %q", name, resourceGroup)
+			}
+			return fmt.Errorf("retrieving Portal Dashboard %q (Resource Group %q): %+v", name, resourceGroup, err)
+		}
+	} else {
+		dashboards := make([]portal.Dashboard, 0)
+
+		iterator, err := client.ListByResourceGroupComplete(ctx, resourceGroup)
+		if err != nil {
+			if utils.ResponseWasNotFound(iterator.Response().Response) {
+				return fmt.Errorf("no Portal Dashboards were found for Resource Group %q", resourceGroup)
+			}
+			return fmt.Errorf("getting list of Portal Dashboards (Resource Group %q): %+v", resourceGroup, err)
+		}
+
+		for iterator.NotDone() {
+			dashboard = iterator.Value()
+
+			found := false
+			for k, v := range dashboard.Tags {
+				if k == "hidden-title" && *v == displayName {
+					found = true
+					break
+				}
+			}
+
+			if found {
+				dashboards = append(dashboards, dashboard)
+			}
+			if err := iterator.NextWithContext(ctx); err != nil {
+				return err
+			}
+		}
+
+		if 1 > len(dashboards) {
+			return fmt.Errorf("no Portal Dashboards were found for Resource Group %q", resourceGroup)
+		}
+
+		if len(dashboards) > 1 {
+			return fmt.Errorf("multiple Portal Dashboards were found for Resource Group %q", resourceGroup)
 		}
-		return fmt.Errorf("retrieving Portal Dashboard %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
+
+		dashboard = dashboards[0]
+	}
+
+	if dashboard.Name == nil {
+		return fmt.Errorf("portal Dashboard name is empty in Resource Group %s", resourceGroup)
 	}
 
+	id := parse.NewDashboardID(subscriptionId, resourceGroup, name)
```

Comment:
As `resourceGroup` is required property, so it won't be empty.

For `name`, while filtering by `display_name`, `name` would be empty. So I updated code to use `*dashboard.Name` since this property always has value when the resource is retrieved.